kazuki-toolkit/src/main/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/Dtg.kt

```diff
@@ -35,203 +37,138 @@ class DtgProperties(private val dtg: Dtg) {
-        post = { duration, result -> result.functions.subtractDuration(duration) == dtg }
+    private val localDateTime by lazy { dtg.toLocalDateTime() }
+
+    val isEarlierThan = function(
+        command = { other: Dtg -> localDateTime < other.toLocalDateTime() },
+        post = { other, result ->
+            result iff (dtg.date.functions.isEarlierThan(other.date) || (
+                    dtg.date == other.date && dtg.time.functions.isEarlierThan(other.time)))
+        }
+    )
+
+    val addDuration = function(
+        command = { duration: Duration ->
+            dtg.toLocalDateTime().plusNanos(duration.milliseconds.toLong() * 1_000_000).toDtg()
+        },
+        post = { duration, result -> DtgUtilities.durationBetween(dtg, result) == duration }
     )
 
     val subtractDuration: (Duration) -> Dtg = function(
-        command = { duration -> dtg.properties.durationSinceFirstDtg.functions.subtractDuration(duration).properties.dtgAfterFirstDtg },
+        command = { duration ->
+            dtg.toLocalDateTime().minusNanos(duration.milliseconds.toLong() * 1_000_000).toDtg()
+        },
         pre = { duration -> duration <= dtg.properties.durationSinceFirstDtg },
-//          post = { duration, result -> result.functions.addDuration(duration) == dtg }
-//          This post condition uses a function whose post condition uses this function as a post condition.
-//          If not commented, the two functions will recur until a stack overflow error occurs.
-//          However, it is still a valid post condition so is left here for completeness.
+        post = { duration, result -> DtgUtilities.durationBetween(result, dtg) == duration }
     )
 
-
-    val withinDurationOfDtg: (Duration, Dtg) -> bool = function(
-        command = { duration, targetDtg ->
-            if (duration.milliseconds == 0uL) {
-                dtg == targetDtg
-            } else {
-                dtg.functions.inInterval(
-                    mk_Interval(
-                        targetDtg.functions.subtractDuration(duration),
-                        targetDtg.functions.addDuration(duration)
-                    )
-                )
-            }
+    val withinDurationOfDtg = function(
+        command = { duration: Duration, other: Dtg ->
+            dtg.functions.inRange(other.functions.subtractDuration(duration), other.functions.addDuration(duration))
         },
-        post = { duration, targetDtg, result ->
-            if (duration == NoDuration) {
-                (dtg == targetDtg) == result
-            } else {
-                (targetDtg.functions.subtractDuration(duration) <= dtg && dtg < targetDtg.functions.addDuration(
-                    duration
-                )) == result
-            }
+        post = { duration, other, result ->
+            result iff dtg.functions.inRange(
+                other.functions.subtractDuration(duration),
+                other.functions.addDuration(duration)
```

Comment:
⛏️  What's the point of this being the exact same as the command above? A better postcondition would be something like
```
DtgUtilities.durationBetween(dtg, other) < duration
```